{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scRNA-seq Imputation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data Availability Statement\n",
    "Tabula Muris data\n",
    "\n",
    "Smart-seq2 https://doi.org/10.6084/m9.figshare.5715040.v1 ( Consortium, The Tabula Muris, 2017a).\n",
    "\n",
    "10X Chromium https://doi.org/10.6084/m9.figshare.5715040.v1 ( Consortium, The Tabula Muris, 2017b).\n",
    "\n",
    "R packages\n",
    "\n",
    "MAGIC: Rmagic (v0.1.0) https://github.com/KrishnaswamyLab/MAGIC\n",
    "\n",
    "DrImpute: DrImpute (v1.0) https://github.com/ikwak2/DrImpute\n",
    "\n",
    "scImpute: scImpute(v0.0.8) https://github.com/Vivianstats/scImpute\n",
    "\n",
    "SAVER: SAVER(v1.0.0) https://github.com/mohuangx/SAVER\n",
    "\n",
    "Knn-smooth: knn_smooth.R (Version 2) https://github.com/yanailab/knn-smoothing\n",
    "\n",
    "Scater: scater(v1.6.3) : https://www.bioconductor.org/packages/release/bioc/html/scater.html\n",
    "\n",
    "Splatter: splatter(v1.2.2) : https://bioconductor.org/packages/release/bioc/html/splatter.html\n",
    "\n",
    "Permute: permute(v0.9-4) : https://cran.r-project.org/web/packages/permute/index.html\n",
    "\n",
    "Python/anaconda packages:\n",
    "\n",
    "Dca : dca(v0.2.2): https://github.com/theislab/dca\n",
    "\n",
    "Custom scripts: https://github.com/tallulandrews/F1000Imputation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Denomising scRNA-seq Data with DCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Running the dca package on definite endoderm cells (DECs)\n",
    "\n",
    "The single cell expression data of DECs are obtained from the GEO website (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75748). The dataset includes the RNA expression count data of 1018 single cells from snapshot progenitors.\n",
    "\n",
    "The paper for analysis of the dataset: Single-cell RNA-seq reveals novel regulators of human embryonic stem cell differentiation to definitive endoderm. https://pubmed.ncbi.nlm.nih.gov/27534536/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !python -m dca.__main__ data/endoderm/endoderm.csv data/endoderm/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>H1_Exp1.001</th>\n",
       "      <th>H1_Exp1.002</th>\n",
       "      <th>H1_Exp1.003</th>\n",
       "      <th>H1_Exp1.004</th>\n",
       "      <th>H1_Exp1.006</th>\n",
       "      <th>H1_Exp1.007</th>\n",
       "      <th>H1_Exp1.008</th>\n",
       "      <th>H1_Exp1.009</th>\n",
       "      <th>H1_Exp1.010</th>\n",
       "      <th>...</th>\n",
       "      <th>TB_Batch2.135</th>\n",
       "      <th>TB_Batch2.136</th>\n",
       "      <th>TB_Batch2.137</th>\n",
       "      <th>TB_Batch2.138</th>\n",
       "      <th>TB_Batch2.139</th>\n",
       "      <th>TB_Batch2.140</th>\n",
       "      <th>TB_Batch2.141</th>\n",
       "      <th>TB_Batch2.142</th>\n",
       "      <th>TB_Batch2.143</th>\n",
       "      <th>TB_Batch2.144</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MKL2</td>\n",
       "      <td>10</td>\n",
       "      <td>162</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "      <td>...</td>\n",
       "      <td>364</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1127</td>\n",
       "      <td>2119</td>\n",
       "      <td>5</td>\n",
       "      <td>500</td>\n",
       "      <td>18</td>\n",
       "      <td>472</td>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD109</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>166</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>53</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>11</td>\n",
       "      <td>48</td>\n",
       "      <td>23</td>\n",
       "      <td>362</td>\n",
       "      <td>22</td>\n",
       "      <td>36</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABTB1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAST2</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>175</td>\n",
       "      <td>41</td>\n",
       "      <td>32</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>206</td>\n",
       "      <td>43</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KAT5</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>52</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>577</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1019 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  H1_Exp1.001  H1_Exp1.002  H1_Exp1.003  H1_Exp1.004  H1_Exp1.006  \\\n",
       "0       MKL2           10          162            3           42            0   \n",
       "1      CD109            6            2          166            9            7   \n",
       "2      ABTB1            0           28            0            1            0   \n",
       "3      MAST2            0          133           41            0            0   \n",
       "4       KAT5            0            7           52           20            0   \n",
       "\n",
       "   H1_Exp1.007  H1_Exp1.008  H1_Exp1.009  H1_Exp1.010  ...  TB_Batch2.135  \\\n",
       "0            2           18            0          182  ...            364   \n",
       "1           53            4           64           29  ...             15   \n",
       "2            9            0            0            0  ...              0   \n",
       "3            2            0            0            0  ...            175   \n",
       "4            6            0            0          103  ...              0   \n",
       "\n",
       "   TB_Batch2.136  TB_Batch2.137  TB_Batch2.138  TB_Batch2.139  TB_Batch2.140  \\\n",
       "0              1             21           1127           2119              5   \n",
       "1             38             38             11             48             23   \n",
       "2              0              0              0              0              0   \n",
       "3             41             32              3              6            206   \n",
       "4            577              0              3              2              0   \n",
       "\n",
       "   TB_Batch2.141  TB_Batch2.142  TB_Batch2.143  TB_Batch2.144  \n",
       "0            500             18            472            350  \n",
       "1            362             22             36             25  \n",
       "2              0              3             39              0  \n",
       "3             43              2              1             99  \n",
       "4             56              2              0              0  \n",
       "\n",
       "[5 rows x 1019 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "cur_dir = os.getcwd()\n",
    "file_path = cur_dir + '/data/endoderm/endoderm.csv'\n",
    "endoderm = pd.read_csv(file_path)\n",
    "endoderm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from keras.layers import Dense, Input\n",
    "from keras.models import Model\n",
    "from dca.layers import ColwiseMultLayer\n",
    "\n",
    "input_size = adata.n_vars\n",
    "output_size = input_size\n",
    "hid_size = 32\n",
    "extra_models = {}\n",
    "\n",
    "def build(adata):\n",
    "    input_layer = Input(shape=(input_size,), name='count')\n",
    "    sf_layer = Input(shape=(1,), name='size_factors')\n",
    "\n",
    "    decoder_output = Dense(hid_size, activation=None, name='center')(input_layer)\n",
    "    mean = Dense(output_size, name='mean')(decoder_output)\n",
    "    output = ColwiseMultLayer([mean, sf_layer])\n",
    "\n",
    "    extra_models['mean_norm'] = Model(inputs=input_layer, outputs=mean)\n",
    "    extra_models['decoded'] = Model(inputs=input_layer, outputs=decoder_output)\n",
    "    \n",
    "    model = Model(inputs=[input_layer, sf_layer], outputs=output)\n",
    " \n",
    "    encoder = Model(inputs=model.input,\n",
    "                    outputs=model.get_layer('center').output)\n",
    "\n",
    "\n",
    "    print('dca: Calculating low dimensional representations...')\n",
    "\n",
    "    adata.obsm['X_dca'] = encoder.predict({'count': adata.X,\n",
    "                                           'size_factors': adata.obs.size_factors})        \n",
    "    print('dca: Calculating reconstructions...')\n",
    "\n",
    "    adata.X = model.predict({'count': adata.X,\n",
    "                             'size_factors': adata.obs.size_factors})\n",
    "    \n",
    "    return adata\n",
    "\n",
    "adata = build(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## scDMFK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /Users/yufeideng/Documents/GitHub/bis687-project/sc-env/lib/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "non-resource variables are not supported in the long term\n",
      "Using TensorFlow backend.\n",
      "begin training\n",
      "Successfully preprocessed 1018 genes and 19097 cells.\n",
      "begin the pretraining\n",
      "2024-04-11 14:40:22.825348: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-04-11 14:40:22.853416: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)\n",
      "[[ -7.61721039  -4.44107199  -5.94987202 ...   2.3466146   -3.77788305\n",
      "    1.32349837]\n",
      " [-14.36012173  -1.99698162  -4.53384209 ...  -5.54547119   1.26300061\n",
      "  -18.28737259]\n",
      " [-39.85637283 -16.12322235  -5.64176178 ...   8.47264957  18.71442413\n",
      "  -59.00668335]\n",
      " ...\n",
      " [  2.1164701   -5.86501551  -5.91279221 ...  -2.14557815  -4.997962\n",
      "   -2.40645623]\n",
      " [  1.19962418  -4.56137419  -4.79713631 ...  -1.66259587  -7.04190731\n",
      "    1.62374187]\n",
      " [-21.29205894 -25.72610855  13.40306664 ...  -3.01693392  15.3586092\n",
      "  -42.71074677]]\n",
      "scDMFK: Saving output(s)...\n"
     ]
    }
   ],
   "source": [
    "!python scDMFK/run.py --dataname \"endoderm/endoderm.csv\" --outputdir \"endoderm/DMFK-results\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/Young'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string = 'data/Young/data.h5'\n",
    "\n",
    "string[:string.rindex(\"/\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scDMFK.utils as utils\n",
    "import numpy as np\n",
    "import h5py\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "\n",
    "def read_clean(data):\n",
    "    assert isinstance(data, np.ndarray)\n",
    "    if data.dtype.type is np.bytes_:\n",
    "        data = utils.decode(data)\n",
    "    if data.size == 1:\n",
    "        data = data.flat[0]\n",
    "    return data\n",
    "\n",
    "def dict_from_group(group):\n",
    "    assert isinstance(group, h5py.Group)\n",
    "    d = utils.dotdict()\n",
    "    for key in group:\n",
    "        if isinstance(group[key], h5py.Group):\n",
    "            value = dict_from_group(group[key])\n",
    "        else:\n",
    "            value = read_clean(group[key][...])\n",
    "        d[key] = value\n",
    "    return d\n",
    "\n",
    "def read_data(filename, sparsify=False, skip_exprs=False):\n",
    "    with h5py.File(filename, \"r\") as f:\n",
    "        obs = pd.DataFrame(dict_from_group(f[\"obs\"]), index=utils.decode(f[\"obs_names\"][...]))\n",
    "        var = pd.DataFrame(dict_from_group(f[\"var\"]), index=utils.decode(f[\"var_names\"][...]))\n",
    "        uns = dict_from_group(f[\"uns\"])\n",
    "        if not skip_exprs:\n",
    "            exprs_handle = f[\"exprs\"]\n",
    "            if isinstance(exprs_handle, h5py.Group):\n",
    "                mat = sp.sparse.csr_matrix((exprs_handle[\"data\"][...], exprs_handle[\"indices\"][...],\n",
    "                                               exprs_handle[\"indptr\"][...]), shape=exprs_handle[\"shape\"][...])\n",
    "            else:\n",
    "                mat = exprs_handle[...].astype(np.float32)\n",
    "                if sparsify:\n",
    "                    mat = sp.sparse.csr_matrix(mat)\n",
    "        else:\n",
    "            mat = sp.sparse.csr_matrix((obs.shape[0], var.shape[0]))\n",
    "    return mat, obs, var, uns\n",
    "\n",
    "def prepro(filename):\n",
    "    data_path = \"data/\" + filename\n",
    "\n",
    "    try:\n",
    "        adata = sc.read(data_path, first_column_names=True)\n",
    "        adata.obs['Group'] = None\n",
    "    except:\n",
    "        mat, obs, var, uns = read_data(data_path, sparsify=False, skip_exprs=False)\n",
    "        if isinstance(mat, np.ndarray):\n",
    "            X = np.array(mat)\n",
    "        else:\n",
    "            X = np.array(mat.toarray())\n",
    "        cell_name = np.array(obs[\"cell_type1\"])\n",
    "        cell_type, cell_label = np.unique(cell_name, return_inverse=True)\n",
    "\n",
    "        X = np.ceil(X).astype(np.int)\n",
    "        adata = sc.AnnData(X)\n",
    "        adata.obs['Group'] = cell_label\n",
    "\n",
    "    print('Successfully preprocessed {} genes and {} cells.'.format(adata.n_vars, adata.n_obs))\n",
    "    return adata\n",
    "\n",
    "\n",
    "def normalize(adata, highly_genes = None, size_factors=True, normalize_input=True, logtrans_input=True):\n",
    "    sc.pp.filter_genes(adata, min_counts=1)\n",
    "    sc.pp.filter_cells(adata, min_counts=1)\n",
    "    if size_factors or normalize_input or logtrans_input:\n",
    "        adata.raw = adata.copy()\n",
    "    else:\n",
    "        adata.raw = adata\n",
    "\n",
    "    if size_factors:\n",
    "        sc.pp.normalize_per_cell(adata)\n",
    "        adata.obs['size_factors'] = adata.obs.n_counts / np.median(adata.obs.n_counts)\n",
    "    else:\n",
    "        adata.obs['size_factors'] = 1.0\n",
    "\n",
    "    if logtrans_input:\n",
    "        sc.pp.log1p(adata)\n",
    "\n",
    "    if highly_genes != None:\n",
    "        sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5, n_top_genes = highly_genes, subset=True)\n",
    "\n",
    "    if normalize_input:\n",
    "        sc.pp.scale(adata)\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1008</th>\n",
       "      <th>1009</th>\n",
       "      <th>1010</th>\n",
       "      <th>1011</th>\n",
       "      <th>1012</th>\n",
       "      <th>1013</th>\n",
       "      <th>1014</th>\n",
       "      <th>1015</th>\n",
       "      <th>1016</th>\n",
       "      <th>1017</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.069419</td>\n",
       "      <td>1.468012</td>\n",
       "      <td>-0.502400</td>\n",
       "      <td>0.907161</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>-0.577798</td>\n",
       "      <td>0.554079</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>1.470833</td>\n",
       "      <td>-0.649936</td>\n",
       "      <td>...</td>\n",
       "      <td>1.326697</td>\n",
       "      <td>-0.755374</td>\n",
       "      <td>0.076480</td>\n",
       "      <td>2.070367</td>\n",
       "      <td>2.234833</td>\n",
       "      <td>-0.298707</td>\n",
       "      <td>1.671089</td>\n",
       "      <td>0.068661</td>\n",
       "      <td>1.497501</td>\n",
       "      <td>1.442243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.682182</td>\n",
       "      <td>-0.867690</td>\n",
       "      <td>0.790358</td>\n",
       "      <td>-0.569321</td>\n",
       "      <td>-0.612554</td>\n",
       "      <td>0.236100</td>\n",
       "      <td>-0.575316</td>\n",
       "      <td>0.476853</td>\n",
       "      <td>-0.276732</td>\n",
       "      <td>-0.555143</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.566862</td>\n",
       "      <td>-0.224366</td>\n",
       "      <td>-0.264537</td>\n",
       "      <td>-0.505149</td>\n",
       "      <td>-0.084520</td>\n",
       "      <td>-0.272777</td>\n",
       "      <td>0.824979</td>\n",
       "      <td>-0.378489</td>\n",
       "      <td>-0.226257</td>\n",
       "      <td>-0.304873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.963323</td>\n",
       "      <td>2.140748</td>\n",
       "      <td>-0.988221</td>\n",
       "      <td>0.405716</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>1.898989</td>\n",
       "      <td>-0.770293</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>-1.021800</td>\n",
       "      <td>-0.974608</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.956583</td>\n",
       "      <td>-0.901495</td>\n",
       "      <td>-0.967738</td>\n",
       "      <td>-0.829039</td>\n",
       "      <td>-0.846477</td>\n",
       "      <td>-0.813906</td>\n",
       "      <td>-0.867951</td>\n",
       "      <td>0.541623</td>\n",
       "      <td>1.685979</td>\n",
       "      <td>-0.850209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.963323</td>\n",
       "      <td>1.994182</td>\n",
       "      <td>1.503586</td>\n",
       "      <td>-0.954341</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>-0.150747</td>\n",
       "      <td>-0.770293</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>-1.021800</td>\n",
       "      <td>0.364907</td>\n",
       "      <td>...</td>\n",
       "      <td>1.495311</td>\n",
       "      <td>0.874722</td>\n",
       "      <td>0.760815</td>\n",
       "      <td>-0.124056</td>\n",
       "      <td>0.095633</td>\n",
       "      <td>1.853983</td>\n",
       "      <td>1.020409</td>\n",
       "      <td>-0.333062</td>\n",
       "      <td>-0.536624</td>\n",
       "      <td>1.365221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.963323</td>\n",
       "      <td>0.180837</td>\n",
       "      <td>1.480572</td>\n",
       "      <td>0.958142</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>0.288384</td>\n",
       "      <td>-0.770293</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>1.618809</td>\n",
       "      <td>1.739285</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.956583</td>\n",
       "      <td>1.961002</td>\n",
       "      <td>-0.967738</td>\n",
       "      <td>-0.221934</td>\n",
       "      <td>-0.388543</td>\n",
       "      <td>-0.813906</td>\n",
       "      <td>1.015744</td>\n",
       "      <td>-0.416733</td>\n",
       "      <td>-0.880896</td>\n",
       "      <td>-0.850209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18245</th>\n",
       "      <td>0.763035</td>\n",
       "      <td>0.953469</td>\n",
       "      <td>-0.119517</td>\n",
       "      <td>1.065267</td>\n",
       "      <td>1.176381</td>\n",
       "      <td>1.022391</td>\n",
       "      <td>1.427621</td>\n",
       "      <td>1.002929</td>\n",
       "      <td>0.928195</td>\n",
       "      <td>1.266762</td>\n",
       "      <td>...</td>\n",
       "      <td>0.973501</td>\n",
       "      <td>-0.557952</td>\n",
       "      <td>1.223269</td>\n",
       "      <td>1.731255</td>\n",
       "      <td>0.920591</td>\n",
       "      <td>-0.579516</td>\n",
       "      <td>0.281782</td>\n",
       "      <td>-0.620383</td>\n",
       "      <td>0.399395</td>\n",
       "      <td>0.147709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18246</th>\n",
       "      <td>-0.963323</td>\n",
       "      <td>-0.963068</td>\n",
       "      <td>-0.988221</td>\n",
       "      <td>-0.954341</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>-0.963207</td>\n",
       "      <td>-0.770293</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>-1.021800</td>\n",
       "      <td>-0.974608</td>\n",
       "      <td>...</td>\n",
       "      <td>1.960357</td>\n",
       "      <td>1.852234</td>\n",
       "      <td>1.830183</td>\n",
       "      <td>2.251798</td>\n",
       "      <td>1.763214</td>\n",
       "      <td>1.709863</td>\n",
       "      <td>2.262206</td>\n",
       "      <td>-0.875434</td>\n",
       "      <td>0.609419</td>\n",
       "      <td>2.119040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18247</th>\n",
       "      <td>0.721264</td>\n",
       "      <td>1.287970</td>\n",
       "      <td>0.565610</td>\n",
       "      <td>1.260535</td>\n",
       "      <td>-0.451260</td>\n",
       "      <td>0.846659</td>\n",
       "      <td>1.868497</td>\n",
       "      <td>0.597759</td>\n",
       "      <td>0.106420</td>\n",
       "      <td>-0.313491</td>\n",
       "      <td>...</td>\n",
       "      <td>1.345113</td>\n",
       "      <td>0.493171</td>\n",
       "      <td>-0.013193</td>\n",
       "      <td>-0.829039</td>\n",
       "      <td>-0.846477</td>\n",
       "      <td>-0.813906</td>\n",
       "      <td>1.585123</td>\n",
       "      <td>-0.875434</td>\n",
       "      <td>1.402134</td>\n",
       "      <td>-0.850209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18248</th>\n",
       "      <td>0.667025</td>\n",
       "      <td>0.750945</td>\n",
       "      <td>1.359231</td>\n",
       "      <td>1.181274</td>\n",
       "      <td>1.025842</td>\n",
       "      <td>1.324286</td>\n",
       "      <td>0.719553</td>\n",
       "      <td>1.220835</td>\n",
       "      <td>1.090974</td>\n",
       "      <td>0.876491</td>\n",
       "      <td>...</td>\n",
       "      <td>1.145226</td>\n",
       "      <td>1.133138</td>\n",
       "      <td>1.149106</td>\n",
       "      <td>1.105013</td>\n",
       "      <td>1.333495</td>\n",
       "      <td>1.297173</td>\n",
       "      <td>1.085032</td>\n",
       "      <td>0.878802</td>\n",
       "      <td>1.643264</td>\n",
       "      <td>1.311002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18249</th>\n",
       "      <td>-0.963323</td>\n",
       "      <td>-0.963068</td>\n",
       "      <td>-0.988221</td>\n",
       "      <td>0.776132</td>\n",
       "      <td>-0.933539</td>\n",
       "      <td>-0.963207</td>\n",
       "      <td>-0.770293</td>\n",
       "      <td>-0.862384</td>\n",
       "      <td>-1.021800</td>\n",
       "      <td>1.327269</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.956583</td>\n",
       "      <td>-0.901495</td>\n",
       "      <td>-0.967738</td>\n",
       "      <td>-0.829039</td>\n",
       "      <td>-0.846477</td>\n",
       "      <td>-0.813906</td>\n",
       "      <td>-0.867951</td>\n",
       "      <td>-0.875434</td>\n",
       "      <td>-0.880896</td>\n",
       "      <td>-0.850209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18250 rows × 1018 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2         3         4         5         6     \\\n",
       "0      0.069419  1.468012 -0.502400  0.907161 -0.933539 -0.577798  0.554079   \n",
       "1     -0.682182 -0.867690  0.790358 -0.569321 -0.612554  0.236100 -0.575316   \n",
       "2     -0.963323  2.140748 -0.988221  0.405716 -0.933539  1.898989 -0.770293   \n",
       "3     -0.963323  1.994182  1.503586 -0.954341 -0.933539 -0.150747 -0.770293   \n",
       "4     -0.963323  0.180837  1.480572  0.958142 -0.933539  0.288384 -0.770293   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "18245  0.763035  0.953469 -0.119517  1.065267  1.176381  1.022391  1.427621   \n",
       "18246 -0.963323 -0.963068 -0.988221 -0.954341 -0.933539 -0.963207 -0.770293   \n",
       "18247  0.721264  1.287970  0.565610  1.260535 -0.451260  0.846659  1.868497   \n",
       "18248  0.667025  0.750945  1.359231  1.181274  1.025842  1.324286  0.719553   \n",
       "18249 -0.963323 -0.963068 -0.988221  0.776132 -0.933539 -0.963207 -0.770293   \n",
       "\n",
       "           7         8         9     ...      1008      1009      1010  \\\n",
       "0     -0.862384  1.470833 -0.649936  ...  1.326697 -0.755374  0.076480   \n",
       "1      0.476853 -0.276732 -0.555143  ... -0.566862 -0.224366 -0.264537   \n",
       "2     -0.862384 -1.021800 -0.974608  ... -0.956583 -0.901495 -0.967738   \n",
       "3     -0.862384 -1.021800  0.364907  ...  1.495311  0.874722  0.760815   \n",
       "4     -0.862384  1.618809  1.739285  ... -0.956583  1.961002 -0.967738   \n",
       "...         ...       ...       ...  ...       ...       ...       ...   \n",
       "18245  1.002929  0.928195  1.266762  ...  0.973501 -0.557952  1.223269   \n",
       "18246 -0.862384 -1.021800 -0.974608  ...  1.960357  1.852234  1.830183   \n",
       "18247  0.597759  0.106420 -0.313491  ...  1.345113  0.493171 -0.013193   \n",
       "18248  1.220835  1.090974  0.876491  ...  1.145226  1.133138  1.149106   \n",
       "18249 -0.862384 -1.021800  1.327269  ... -0.956583 -0.901495 -0.967738   \n",
       "\n",
       "           1011      1012      1013      1014      1015      1016      1017  \n",
       "0      2.070367  2.234833 -0.298707  1.671089  0.068661  1.497501  1.442243  \n",
       "1     -0.505149 -0.084520 -0.272777  0.824979 -0.378489 -0.226257 -0.304873  \n",
       "2     -0.829039 -0.846477 -0.813906 -0.867951  0.541623  1.685979 -0.850209  \n",
       "3     -0.124056  0.095633  1.853983  1.020409 -0.333062 -0.536624  1.365221  \n",
       "4     -0.221934 -0.388543 -0.813906  1.015744 -0.416733 -0.880896 -0.850209  \n",
       "...         ...       ...       ...       ...       ...       ...       ...  \n",
       "18245  1.731255  0.920591 -0.579516  0.281782 -0.620383  0.399395  0.147709  \n",
       "18246  2.251798  1.763214  1.709863  2.262206 -0.875434  0.609419  2.119040  \n",
       "18247 -0.829039 -0.846477 -0.813906  1.585123 -0.875434  1.402134 -0.850209  \n",
       "18248  1.105013  1.333495  1.297173  1.085032  0.878802  1.643264  1.311002  \n",
       "18249 -0.829039 -0.846477 -0.813906 -0.867951 -0.875434 -0.880896 -0.850209  \n",
       "\n",
       "[18250 rows x 1018 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Endoderm dataset\n",
    "\n",
    "# adata = prepro('endoderm/endoderm.csv')\n",
    "# X = adata.X.astype(np.float32)\n",
    "\n",
    "# adata = normalize(adata, highly_genes=adata.n_vars)\n",
    "# sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5, n_top_genes = 10, subset=True)\n",
    "\n",
    "adata.var.highly_variable.index\n",
    "pd.DataFrame(adata.X[:, adata.var.highly_variable])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5680</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5681</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5682</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5683</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5684</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5685 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Group\n",
       "0         6\n",
       "1         6\n",
       "2         6\n",
       "3         6\n",
       "4         6\n",
       "...     ...\n",
       "5680      0\n",
       "5681      0\n",
       "5682      0\n",
       "5683      0\n",
       "5684      0\n",
       "\n",
       "[5685 rows x 1 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata = prepro('Young/data.h5')\n",
    "# (adata.X != 0).any(axis=0)\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Iterative imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.experimental import enable_iterative_imputer\n",
    "from sklearn.impute import IterativeImputer\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "endoderm = pd.read_csv('data/endoderm/endoderm.csv', index_col=0)\n",
    "endoderm.head()\n",
    "endoderm = endoderm.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4 µs, sys: 1 µs, total: 5 µs\n",
      "Wall time: 9.3 µs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yufeideng/Documents/GitHub/bis687-project/sc-env/lib/python3.8/site-packages/sklearn/impute/_base.py:558: UserWarning: Skipping features without any observed values: [    7    16    36    42    55    63    68    96    99   132   145   172\n",
      "   207   210   244   283   290   304   331   334   335   444   450   457\n",
      "   471   495   539   543   572   580   637   669   672   674   679   716\n",
      "   736   767   775   929   936   938  1005  1013  1051  1060  1069  1146\n",
      "  1247  1261  1333  1370  1386  1402  1423  1482  1544  1546  1550  1566\n",
      "  1597  1605  1637  1642  1645  1686  1714  1729  1782  1786  1789  1826\n",
      "  1870  1872  1877  1881  1924  1953  1979  1981  2011  2012  2022  2057\n",
      "  2089  2122  2129  2134  2161  2218  2224  2241  2256  2261  2279  2286\n",
      "  2294  2331  2381  2384  2385  2395  2452  2485  2487  2492  2523  2528\n",
      "  2587  2591  2608  2623  2630  2669  2684  2685  2729  2730  2750  2820\n",
      "  2827  2829  2846  2859  2873  2890  2892  2899  2914  2923  2932  2941\n",
      "  2984  3037  3041  3077  3095  3100  3101  3104  3113  3144  3165  3172\n",
      "  3185  3196  3238  3245  3248  3253  3281  3313  3382  3396  3402  3409\n",
      "  3422  3427  3430  3457  3477  3489  3496  3524  3548  3573  3616  3620\n",
      "  3638  3673  3677  3683  3703  3711  3724  3798  3817  3860  3865  3908\n",
      "  3914  3927  3929  3934  3951  3966  3990  4019  4027  4054  4067  4133\n",
      "  4150  4153  4159  4200  4225  4235  4243  4308  4383  4393  4436  4474\n",
      "  4502  4508  4517  4526  4529  4531  4535  4544  4547  4550  4579  4601\n",
      "  4608  4609  4618  4649  4721  4740  4745  4748  4775  4801  4803  4813\n",
      "  4832  4836  4848  4870  4889  4923  5000  5060  5071  5079  5119  5145\n",
      "  5199  5242  5268  5274  5295  5350  5356  5380  5440  5469  5474  5484\n",
      "  5486  5488  5497  5513  5549  5600  5617  5652  5664  5676  5690  5710\n",
      "  5740  5748  5773  5798  5831  5844  5856  5903  5908  5910  5949  5950\n",
      "  5969  5978  5983  5984  5986  6046  6064  6072  6099  6115  6145  6184\n",
      "  6214  6241  6249  6283  6326  6339  6356  6371  6383  6462  6475  6477\n",
      "  6497  6498  6556  6582  6609  6617  6651  6668  6692  6708  6711  6713\n",
      "  6715  6760  6762  6801  6812  6817  6829  6858  6907  6919  6970  6984\n",
      "  7008  7156  7160  7215  7216  7242  7253  7334  7367  7369  7373  7376\n",
      "  7381  7384  7393  7396  7404  7421  7441  7498  7513  7521  7535  7539\n",
      "  7557  7558  7561  7563  7615  7619  7725  7735  7808  7858  7874  7890\n",
      "  7932  7989  8022  8043  8051  8064  8101  8171  8203  8208  8235  8236\n",
      "  8289  8290  8295  8311  8319  8340  8358  8362  8383  8389  8431  8438\n",
      "  8480  8493  8503  8508  8527  8561  8607  8617  8625  8636  8648  8649\n",
      "  8651  8653  8672  8701  8728  8760  8781  8807  8822  8828  8902  8962\n",
      "  8965  8991  9037  9098  9126  9137  9139  9155  9167  9170  9199  9216\n",
      "  9249  9275  9283  9317  9332  9348  9363  9425  9435  9452  9477  9487\n",
      "  9494  9515  9520  9523  9568  9594  9625  9632  9651  9664  9677  9701\n",
      "  9715  9734  9754  9761  9763  9784  9788  9789  9790  9825  9827  9863\n",
      "  9864  9868  9872  9913  9936  9963  9966 10014 10019 10028 10041 10045\n",
      " 10062 10113 10117 10121 10127 10131 10145 10180 10215 10238 10240 10243\n",
      " 10287 10305 10307 10369 10383 10393 10442 10448 10452 10476 10492 10505\n",
      " 10527 10564 10584 10585 10593 10594 10630 10634 10654 10662 10679 10687\n",
      " 10694 10700 10715 10724 10738 10741 10743 10745 10757 10765 10766 10787\n",
      " 10860 10891 10899 10960 10989 11003 11071 11074 11076 11101 11115 11149\n",
      " 11154 11165 11167 11177 11185 11192 11203 11212 11216 11273 11282 11315\n",
      " 11355 11358 11378 11399 11473 11487 11528 11547 11575 11598 11612 11627\n",
      " 11632 11658 11693 11707 11711 11770 11779 11793 11833 11853 11892 11950\n",
      " 11961 12084 12092 12094 12099 12106 12138 12160 12161 12183 12221 12232\n",
      " 12245 12269 12283 12289 12343 12349 12364 12369 12377 12388 12406 12421\n",
      " 12439 12476 12477 12487 12522 12535 12545 12554 12555 12660 12704 12720\n",
      " 12722 12727 12739 12747 12750 12765 12785 12801 12819 12852 12853 12881\n",
      " 12904 12923 12952 12992 13001 13041 13051 13107 13144 13168 13172 13174\n",
      " 13178 13180 13184 13191 13213 13235 13311 13320 13375 13406 13416 13417\n",
      " 13422 13444 13482 13510 13515 13538 13554 13571 13576 13620 13650 13701\n",
      " 13724 13765 13785 13787 13803 13805 13807 13812 13849 13865 13878 13879\n",
      " 13911 13921 13982 14000 14010 14040 14041 14042 14070 14095 14101 14106\n",
      " 14108 14113 14125 14142 14157 14198 14204 14232 14237 14262 14302 14330\n",
      " 14331 14340 14423 14436 14489 14501 14508 14517 14518 14529 14543 14544\n",
      " 14546 14554 14575 14589 14597 14607 14631 14651 14670 14672 14673 14680\n",
      " 14701 14706 14723 14729 14741 14746 14778 14799 14808 14850 14881 14888\n",
      " 14907 14923 14959 14965 14973 14990 14993 15002 15014 15018 15020 15051\n",
      " 15066 15081 15112 15120 15131 15151 15153 15239 15261 15290 15314 15320\n",
      " 15324 15337 15358 15360 15442 15461 15475 15477 15494 15508 15521 15568\n",
      " 15569 15605 15608 15610 15619 15627 15629 15637 15668 15674 15683 15720\n",
      " 15757 15758 15798 15805 15814 15836 15840 16040 16041 16051 16084 16086\n",
      " 16117 16125 16132 16152 16162 16166 16174 16184 16202 16208 16222 16234\n",
      " 16286 16299 16311 16320 16339 16361 16370 16383 16402 16419 16441 16494\n",
      " 16501 16514 16515 16609 16612 16622 16667 16726 16731 16740 16745 16753\n",
      " 16780 16782 16790 16857 16859 16905 16946 16951 17048 17051 17090 17096\n",
      " 17123 17151 17153 17176 17191 17198 17202 17207 17282 17295 17333 17361\n",
      " 17374 17396 17403 17422 17456 17463 17499 17508 17509 17532 17566 17580\n",
      " 17600 17612 17622 17631 17632 17646 17648 17665 17668 17669 17697 17750\n",
      " 17754 17756 17761 17762 17785 17789 17793 17863 17879 17880 17888 17889\n",
      " 17975 17995 18003 18009 18030 18043 18065 18117 18137 18149 18160 18180\n",
      " 18218 18224 18327 18366 18380 18410 18497 18511 18535 18582 18610 18627\n",
      " 18688 18736 18739 18780 18786 18790 18791 18792 18804 18810 18816 18828\n",
      " 18867 18881 18929 18943 18948 18957 18973 18986 18994 19018 19052 19055\n",
      " 19071 19077]. At least one non-missing value is needed for imputation with strategy='mean'.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "data = endoderm.iloc[:50, :50]\n",
    "# train test split\n",
    "from sklearn.model_selection import train_test_split\n",
    "X_train, X_test = train_test_split(endoderm, test_size=0.2, random_state=42)\n",
    "\n",
    "# initialization strategy: mean\n",
    "%time\n",
    "imp_mean = IterativeImputer(random_state=0, missing_values = 0)\n",
    "imp_mean.fit(X_train)\n",
    "transformed = imp_mean.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dca-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
